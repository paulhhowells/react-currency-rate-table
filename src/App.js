import React from 'react';
import './App.css';
import CurrencyRateTable from './components/CurrencyRateTable/';

// TODO: think about { Component } from React.

export default () => (
  <div>
    <CurrencyRateTable />
  </div>
);