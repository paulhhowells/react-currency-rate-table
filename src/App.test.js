import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { mount, configure } from 'enzyme';
import 'jest-enzyme';
import rates from '../public/api/ratesTest.json';
import App from './App';

configure({
    adapter: new Adapter()
});

global.fetch = () => Promise.resolve({
    json: () => Promise.resolve(rates)
});

// based on https://blog.pragmatists.com/genuine-guide-to-testing-react-redux-applications-6f3265c11f63
const flushAllPromises = () => new Promise(resolve => setImmediate(resolve));

export const flushRequestsAndUpdate = async enzymeWrapper => {
    await flushAllPromises()
    enzymeWrapper.update();
}

it('renders without crashing', () => {
    mount(<App />);
});

it('renders loading text initially', async () => {
    const app = mount(<App />);
    expect(app.find('div')).toHaveText('Loading...');
});

it('renders a table after data load', async () => {
    const app = mount(<App />);
    expect(app.find('div')).toHaveText('Loading...');
    await flushRequestsAndUpdate(app);
    expect(app.find('table')).toExist();
});

it('renders rows with currency name as key', async () => {
    const app = mount( <App /> );
    await flushRequestsAndUpdate(app);

    expect(app.find('table tbody tr').at(0).key()).toEqual('AED');
    expect(app.find('table tbody tr').at(1).key()).toEqual('AFN');
});

it('renders columns with currency name concatenated with column index as key', async () => {
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);

    expect(app.find('table tbody tr').at(0).childAt(0).key()).toEqual('AED0');
    expect(app.find('table tbody tr').at(0).childAt(1).key()).toEqual('AED1');
});

it('renders table that is initially sorted by currency ascending', async () => {
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);
    expect(app.find('table')).toMatchSnapshot();
});

it('can sort first column descending', async () => {
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);
    app.find('table thead tr th').at(0).simulate('click');
    expect(app.find('table')).toMatchSnapshot();
});

it('can sort by second column ascending', async () => {
    const app = mount( <App /> );
    await flushRequestsAndUpdate(app);
    app.find('table thead tr th').at(1).simulate('click');
    expect(app.find('table')).toMatchSnapshot();
});

it('can sort by second column descending', async () => {
    const app = mount(<App />);
    await flushRequestsAndUpdate(app);
    app.find('table thead tr th').at(1).simulate('click');
    app.find('table thead tr th').at(1).simulate('click');
    expect(app.find('table')).toMatchSnapshot();
});