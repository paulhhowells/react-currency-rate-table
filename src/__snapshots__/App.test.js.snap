// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can sort by second column ascending 1`] = `
<table>
  <thead>
    <tr>
      <th
        onClick={[Function]}
      >
        Currency
      </th>
      <th
        onClick={[Function]}
      >
        Rate
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="AED"
    >
      <td
        key="AED0"
      >
        AED
      </td>
      <td
        key="AED1"
      >
        4.271175
      </td>
    </tr>
    <tr
      key="EGP"
    >
      <td
        key="EGP0"
      >
        EGP
      </td>
      <td
        key="EGP1"
      >
        20.790119
      </td>
    </tr>
    <tr
      key="AFN"
    >
      <td
        key="AFN0"
      >
        AFN
      </td>
      <td
        key="AFN1"
      >
        84.428054
      </td>
    </tr>
    <tr
      key="ZWL"
    >
      <td
        key="ZWL0"
      >
        ZWL
      </td>
      <td
        key="ZWL1"
      >
        374.873376
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`can sort by second column descending 1`] = `
<table>
  <thead>
    <tr>
      <th
        onClick={[Function]}
      >
        Currency
      </th>
      <th
        onClick={[Function]}
      >
        Rate
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="ZWL"
    >
      <td
        key="ZWL0"
      >
        ZWL
      </td>
      <td
        key="ZWL1"
      >
        374.873376
      </td>
    </tr>
    <tr
      key="AFN"
    >
      <td
        key="AFN0"
      >
        AFN
      </td>
      <td
        key="AFN1"
      >
        84.428054
      </td>
    </tr>
    <tr
      key="EGP"
    >
      <td
        key="EGP0"
      >
        EGP
      </td>
      <td
        key="EGP1"
      >
        20.790119
      </td>
    </tr>
    <tr
      key="AED"
    >
      <td
        key="AED0"
      >
        AED
      </td>
      <td
        key="AED1"
      >
        4.271175
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`can sort first column descending 1`] = `
<table>
  <thead>
    <tr>
      <th
        onClick={[Function]}
      >
        Currency
      </th>
      <th
        onClick={[Function]}
      >
        Rate
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="ZWL"
    >
      <td
        key="ZWL0"
      >
        ZWL
      </td>
      <td
        key="ZWL1"
      >
        374.873376
      </td>
    </tr>
    <tr
      key="EGP"
    >
      <td
        key="EGP0"
      >
        EGP
      </td>
      <td
        key="EGP1"
      >
        20.790119
      </td>
    </tr>
    <tr
      key="AFN"
    >
      <td
        key="AFN0"
      >
        AFN
      </td>
      <td
        key="AFN1"
      >
        84.428054
      </td>
    </tr>
    <tr
      key="AED"
    >
      <td
        key="AED0"
      >
        AED
      </td>
      <td
        key="AED1"
      >
        4.271175
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`renders table that is initially sorted by currency ascending 1`] = `
<table>
  <thead>
    <tr>
      <th
        onClick={[Function]}
      >
        Currency
      </th>
      <th
        onClick={[Function]}
      >
        Rate
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="AED"
    >
      <td
        key="AED0"
      >
        AED
      </td>
      <td
        key="AED1"
      >
        4.271175
      </td>
    </tr>
    <tr
      key="AFN"
    >
      <td
        key="AFN0"
      >
        AFN
      </td>
      <td
        key="AFN1"
      >
        84.428054
      </td>
    </tr>
    <tr
      key="EGP"
    >
      <td
        key="EGP0"
      >
        EGP
      </td>
      <td
        key="EGP1"
      >
        20.790119
      </td>
    </tr>
    <tr
      key="ZWL"
    >
      <td
        key="ZWL0"
      >
        ZWL
      </td>
      <td
        key="ZWL1"
      >
        374.873376
      </td>
    </tr>
  </tbody>
</table>
`;
